### PR TITLE
[AArch64] Refactor creation of a shuffle mask for TBL (NFC)

### PR DESCRIPTION
**Pull request by momchil-velikov**
... in preparation for https://github.com/llvm/llvm-project/pull/92528

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 1]
```diff
@@ -15710,48 +15710,51 @@ bool AArch64TargetLowering::shouldSinkOperands(
...
+
+  Mask.clear();
+  Mask.resize(MaskLen, NumElts);
+
+  unsigned SrcIndex = 0;
+  for (unsigned I = 0; I < MaskLen; I += Factor)
+    Mask[I] = SrcIndex++;
+
+  if (!IsLittleEndian)
+    std::rotate(Mask.rbegin(), Mask.rbegin() + Factor - 1, Mask.rend());
```
Comment (momchil-velikov) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:15733]: Eh, this could be a rotate by 1 in the opposite direction, will fix later.

File: llvm/lib/Target/AArch64/AArch64InstrInfo.td [hunk 2]
```diff
@@ -735,6 +735,12 @@ def AArch64rev64 : SDNode<"AArch64ISD::REV64", SDT_AArch64UnaryVec>;
 def AArch64ext : SDNode<"AArch64ISD::EXT", SDT_AArch64ExtVec>;
 
 def AArch64vashr : SDNode<"AArch64ISD::VASHR", SDT_AArch64vshift>;
+
+def AArch64vashr_exact : PatFrag<(ops          node:$lhs, node:$rhs),
```
Comment (fhahn) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.td:739]: looks like some changes from the follow-up commit made it in this PR?

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 3]
```diff
@@ -15916,21 +15919,30 @@ bool AArch64TargetLowering::optimizeExtendOrTruncateConversion(
 
       DstTy = TruncDstType;
     }
-
-    return createTblShuffleForZExt(ZExt, DstTy, Subtarget->isLittleEndian());
+    IRBuilder<> Builder(ZExt);
+    Value *Result = createTblShuffleForZExt(
```
Comment (fhahn) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:15923]: could `createTblShuffleForZExt` still take care of replacing and erasing?

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 4]
```diff
@@ -15710,48 +15710,51 @@ bool AArch64TargetLowering::shouldSinkOperands(
...
-  auto *SrcTy = cast<FixedVectorType>(Op->getType());
-  auto SrcWidth = cast<IntegerType>(SrcTy->getElementType())->getBitWidth();
-  auto DstWidth = cast<IntegerType>(DstTy->getElementType())->getBitWidth();
+static bool createTblShuffleMask(unsigned SrcWidth, unsigned DstWidth,
+                                 unsigned NumElts, bool IsLittleEndian,
+                                 SmallVectorImpl<int> &Mask) {
   if (DstWidth % 8 != 0 || DstWidth <= 16 || DstWidth >= 64)
     return false;
 
-  assert(DstWidth % SrcWidth == 0 &&
```
Comment (fhahn) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:15722]: Is the assert replaced by an `if` because it may trigger, possibly on an untested path?

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 5]
```diff
@@ -14328,7 +14328,7 @@ SDValue AArch64TargetLowering::LowerVectorSRA_SRL_SHL(SDValue Op,
       unsigned Opc =
           (Op.getOpcode() == ISD::SRA) ? AArch64ISD::VASHR : AArch64ISD::VLSHR;
       return DAG.getNode(Opc, DL, VT, Op.getOperand(0),
-                         DAG.getConstant(Cnt, DL, MVT::i32));
```
Comment (fhahn) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:14331]: unrelated?

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp (hunk 3 above)
Comment (momchil-velikov) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:15923]: I pulled the erasing out of `createTblShuffleForZExt` to avoid the situation down below in the `UIToFPInst` case where a `zext` is created and immediately deleted.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp (hunk 5 above)
Comment (momchil-velikov) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:14331]: This one is part of another pull request. The dependencies are like this:

#92922 <--- #92528 ---> #92529

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp (hunk 4 above)
Comment (momchil-velikov) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:15722]: No, it just felt more like the conditions above rather than a compiler error / invariant violation that warrants an abort.

**Comment by momchil-velikov** [comment]
Ping?

**Review by davemgreen — APPROVED** [review]
I believe this is fine too, if there are no other comments.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp (hunk 4 above)
Comment (fhahn) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp]: Right, would be good to have a test case if this can happen

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp (hunk 3 above)
Comment (fhahn) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp]: Personally I'd go with simpler code, rather than avoiding creating a new ZExt which then gets removed, but don;t feel too strongly about it

**Review by fhahn — APPROVED** [review]
LGTM, would be good to have a test case & remove the NFC tag if assert is replaced by early exit

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp (hunk 4 above)
Comment (momchil-velikov) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp]: Currently it can't happen, all the callers are with `i8` (i.e. `SrcWidth == 8`)

Comment (fhahn) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp]: Then keeping the code assert would seem appropriate, unless it is changes in the follow-up?
